File: src/controller.rs
#[derive(Debug, Default)]
pub struct Controller {
    pub selected: usize,
}

impl Controller {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn scroll_up(&mut self) {
        self.selected = self.selected.saturating_sub(1);
    }

    pub fn scroll_down(&mut self) {
        self.selected = self.selected.saturating_add(1);
    }

    pub fn page_up(&mut self) {
        self.selected = self.selected.saturating_sub(10);
    }

    pub fn page_down(&mut self) {
        self.selected = self.selected.saturating_add(10);
    }
}

File: src/controller.zig
//! Zig's `+|` / `-|` are built-in saturating-arithmetic operators, so this
//! is a near-literal translation of `saturating_add` / `saturating_sub`.
pub const Controller = struct {
    selected: usize = 0,

    pub fn scrollUp(self: *Controller) void {
        self.selected -|= 1;
    }

    pub fn scrollDown(self: *Controller) void {
        self.selected +|= 1;
    }

    pub fn pageUp(self: *Controller) void {
        self.selected -|= 10;
    }

    pub fn pageDown(self: *Controller) void {
        self.selected +|= 10;
    }
};
